### PR TITLE
[HeteroGraph] Metis partition for hetero

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/partition.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        subg.ndata['part_id'] = F.gather_row(node_part, subg.ndata[NID])
+        if reshuffle:
+            subg.ndata['orig_id'] = F.gather_row(orig_nids, subg.ndata[NID])
+            subg.edata['orig_id'] = F.gather_row(orig_eids, subg.edata[EID])
+
+        if extra_cached_hops >= 1:
+            inner_edge = F.zeros((subg.number_of_edges(),), F.int8, F.cpu())
+            inner_nids = F.nonzero_1d(subg.ndata['inner_node'])
+            # TODO(zhengda) we need to fix utils.toindex() to avoid the dtype cast below.
+            inner_nids = F.astype(inner_nids, F.int64)
```
Comment (zheng-da) [inline python/dgl/partition.py:133]: we should remove the type case.

File: python/dgl/transform.py [hunk 2]
```diff
@@ -18,7 +18,9 @@
 from . import utils
 from .base import EID, NID
 from . import ndarray as nd
-
+from .partition import metis_partition_assignment as hetero_metis_partition_assignment
+from .partition import partition_graph_with_halo as hetero_partition_graph_with_halo
+from .partition import metis_partition as hetero_metis_partition
```
Comment (zheng-da) [inline python/dgl/transform.py:23]: i think we can just remove the old implementations.

File: tests/compute/test_transform.py [hunk 3]
```diff
@@ -1013,10 +1024,11 @@ def test_cast():
     # test_remove_self_loop()
     # test_add_self_loop()
     # test_partition_with_halo()
-    # test_metis_partition()
+    test_metis_partition()
+    test_hetero_metis_partition()
     # test_hetero_linegraph('int32')
     # test_compact()
-    test_to_simple("int32")
+    # test_to_simple("int32")
```
Comment (zheng-da) [inline tests/compute/test_transform.py:1031]: why comment out the tests?

File: tests/compute/test_transform.py [hunk 4]
```diff
@@ -490,6 +490,17 @@ def test_metis_partition():
...
 
+@unittest.skipIf(F._default_context_str == 'gpu', reason="METIS doesn't support GPU")
+def test_hetero_metis_partition():
+    # TODO(zhengda) Metis fails to partition a small graph.
+    g = dgl.DGLGraph(create_large_graph_index(1000), readonly=True)
+    g = dgl.as_heterograph(g)
+    check_metis_partition(g, 0)
+    check_metis_partition(g, 1)
+    check_metis_partition(g, 2)
+    check_metis_partition_with_constraint(g)
```
Comment (zheng-da) [inline tests/compute/test_transform.py:501]: it's better to adapt dgl.distributed.partition_graph and run through all tests in tests/distributed/test_partition.py as well.

File: src/graph/transform/partition_hetero.cc [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    });
+  });
+
+DGL_REGISTER_GLOBAL("partition._CAPI_GetHaloSubgraphInnerNodes_Hetero")
+  .set_body([](DGLArgs args, DGLRetValue *rv) {
+    HeteroSubgraphRef g = args[0];
+    auto gptr = std::dynamic_pointer_cast<HaloHeteroSubgraph>(g.sptr());
+    CHECK(gptr) << "The input graph has to be HaloHeteroSubgraph";
+    *rv = gptr->inner_nodes[0];
+  });
```
Comment (zheng-da) [inline src/graph/transform/partition_hetero.cc:256]: most of the code here is copied from the old code?

File: tests/compute/test_transform.py (hunk 4 above)
Comment (VoVAllen) [reply tests/compute/test_transform.py:501]: Added

File: src/graph/transform/partition_hetero.cc (hunk 5 above)
Comment (VoVAllen) [reply src/graph/transform/partition_hetero.cc:257]: Yes

File: tests/compute/test_transform.py (hunk 3 above)
Comment (VoVAllen) [reply tests/compute/test_transform.py:1031]: It's just for debug usage. Pytest will run all functions with name starting with "test"

File: python/dgl/transform.py (hunk 2 above)
Comment (VoVAllen) [reply python/dgl/transform.py:23]: Preserved for now to debug further end2end examples

**Pull request merged**